src/Dfc.CourseDirectory.WebV2/Features/Courses/DeleteCourseRun.cs

```diff
@@ -0,0 +1,168 @@
+                {
+                    new ValidationFailure(nameof(request.Confirm), "Confirm you want to delete the course")
+                });
+                return new ModelWithErrors<ViewModel>(vm, validationResult);
+            }
+
+            await _cosmosDbQueryDispatcher.ExecuteQuery(new DeleteCourseRunQuery()
+            {
+                CourseId = request.CourseId,
+                CourseRunId = request.CourseRunId,
+                ProviderUkprn = ukprn
+            });
+
+            return new SuccessViewModel()
+            {
+                CourseId = request.CourseId,
+                CourseRunId = request.CourseRunId,
+                CourseName = courseRun.CourseName
+            };
+        }
+
+        private async Task<ViewModel> CreateViewModel(Course course, CourseRun courseRun)
+        {
+            Venue venue = default;
+            if (courseRun.VenueId.HasValue)
+            {
+                venue = await _cosmosDbQueryDispatcher.ExecuteQuery(new GetVenueById() { VenueId = courseRun.VenueId.Value });
+            }
+
+            return new ViewModel()
+            {
+                CourseId = course.Id,
+                CourseRunId = courseRun.Id,
+                CourseName = courseRun.CourseName,
+                DeliveryMode = courseRun.DeliveryMode,
+                FlexibleStartDate = courseRun.FlexibleStartDate,
+                StartDate = courseRun.StartDate,
+                VenueName = venue?.VenueName,
+                YourReference = courseRun.ProviderCourseID
+            };
+        }
+
+        private async Task<(Course Course, CourseRun CourseRun)> GetCourseAndCourseRun(Guid courseId, Guid courseRunId, int ukprn)
+        {
+            (await _cosmosDbQueryDispatcher.ExecuteQuery(new GetCoursesByIds()
+            {
+                CourseIds = new[] { courseId },
+                Ukprn = ukprn
+            })).TryGetValue(courseId, out var course);
+
+            if (course == null)
+            {
+                throw new ResourceDoesNotExistException(ResourceType.Course, courseId);
+            }
+
+            var courseRun = course.CourseRuns.SingleOrDefault(cr => cr.Id == courseRunId);
+
+            if (courseRun == null)
+            {
+                throw new ResourceDoesNotExistException(ResourceType.CourseRun, courseRunId);
+            }
+
+            return (course, courseRun);
+        }
+
+        private async Task<int> GetUkprnForCourse(Guid courseId)
+        {
+            var providerId = await _providerOwnershipCache.GetProviderForCourse(courseId);
+
+            if (!providerId.HasValue)
+            {
+                throw new ResourceDoesNotExistException(ResourceType.Course, courseId);
+            }
+
+            var providerInfo = await _providerInfoCache.GetProviderInfo(providerId.Value);
+            return providerInfo.Ukprn;
+        }
+
+        Guid IRequireUserCanAccessCourse<Request>.GetCourseId(Request request) => request.CourseId;
```

Comment:
For the sake of my understanding, what does this do? I'm guessing access control, but I can't guess how it fits into the flow etc.

Reply:
If a handler implements `IRequireUserCanAccessCourse<TRequest>` then it will get the course authorization rules applied to it (defined by https://github.com/SkillsFundingAgency/dfc-coursedirectory/blob/main/src/Dfc.CourseDirectory.WebV2/Behaviors/RequireUserCanAccessCourseBehavior.cs)